### PR TITLE
3648 - 1

### PR DESCRIPTION
**Pull request by xxkvl**
## Номер
3648

## Номер задания
1

## Ссылка на видео с демо
output

## Комментарии
Хотел сделать то-то, но не успел и не понимаю.

### CODE REVIEW

File: 3648/1/gem_stat.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require 'nokogiri'
+require 'open-uri'
+require 'httparty'
+
+AUTH_TOKEN = '43a345ee370fdfafc836b3534b4311a3a8ee45df'
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: Move token to separate yml file

File: 3648/1/gem_stat.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  attr_reader :gem_name
+  def initialize(gem_name)
+    @gem_name = gem_name
+  end
+
+  def create_stat
+    {
+      gem_name: gem_name,
+      used_by: find_used_by,
+      watch_by: find_wath_by,
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: Spelling mistake. And if you decided to use past tense, use it everywhere. By past tense I mean `watched_by`

File: 3648/1/gem_stat.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      stars: find_stars,
+      forks: find_forks,
+      contributors: find_contributors,
+      open_issues: find_issues
+    }
+  end
+
+  private
+
+  # def gem?
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: Do we need this commented code in here?

File: 3648/1/gem_stat.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  #   url = "https://rubygems.org/api/v1/gems/#{gem_name}.json"
+  #   HTTParty.get(url).parsed_response['name'].nil?
+  # end
+
+  def find_repo
+    # raise StandardError if gem?
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: The same about commented code

File: 3648/1/gem_stat.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    url = "https://#{AUTH_TOKEN}@api.github.com/search/repositories?q=#{gem_name}&per_page=1"
+    HTTParty.get(url).parsed_response['items'][0]['html_url'].to_s
+    # rescue StandardError => e
+    #   puts "ERR: [#{gem_name}] gem could not be found."
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: // Commented code

File: 3648/1/gem_stat.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def downgload_html
+    @downgload_html ||= Nokogiri.HTML(::Kernel.open(find_repo))
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: Use `File.open` instead of Kernel

File: 3648/1/gem_stat.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def downgload_used_html
+    @downgload_used_html ||= Nokogiri.HTML(::Kernel.open("#{find_repo}/network/dependents"))
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: Use `File.open` instead of Kernel

File: 3648/1/gems.yml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+gems:
+  - rails
+  - nokogiri
+  - rspec
```
Comment (andreydro) [inline 3648/1/gems.yml]: Add new line. And setup your text editor for automatic new line adding

File: 3648/1/top_gems.rb [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  end
+  parser.on('-n', '--name=NAME') do |name|
+    options[:word] = name
+  end
+  parser.on('-f', '--file=FILE') do |file|
+    options[:file_name] = file
+  end
+end.parse!
+
+class Output
```
Comment (andreydro) [inline 3648/1/top_gems.rb]: Move `Output` to different file

Comment (andreydro) [inline 3648/1/top_gems.rb]: Separate `private` and `public` methods in your class

**Comment by xxkvl** [comment]
@andreydro i think, all done

File: 3648/1/gem_stat.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  attr_reader :gem_name
+  def initialize(gem_name)
+    @gem_name = gem_name
+  end
+
+  def create_stat
+    {
+      gem_name: gem_name,
+      used_by: find_used_by,
+      watch_by: find_wathed_by,
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: Be attentive to details. You made another spelling mistake and didn't rename field to past tense

Comment (andreydro) [reply 3648/1/gem_stat.rb]: This spelling mistake trailing further in code

File: 3648/1/output.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+  def print_table
+    rows = []
+    select_top_gems.each { |gem| rows << gem.values }
+    puts Terminal::Table.new headings: HEADERS, rows: rows
+  end
+
+  private
+
+  def parse_flag
```
Comment (andreydro) [inline 3648/1/output.rb]: You assign one instance variable to another one. It's useless assignment. If you want just to have different name, then name it right from the first attempt in `initialize`

File: 3648/1/gem_stat.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+      forks: find_forks,
+      contributors: find_contributors,
+      open_issues: find_issues
+    }
+  end
+
+  private
+
+  def find_repo
+    url = "https://#{AUTH_TOKEN}@api.github.com/search/repositories?q=#{gem_name}&per_page=1"
```
Comment (andreydro) [inline 3648/1/gem_stat.rb]: Move logic for API call to separate class (and file). It is better when class does only one particular job. If API changes in future, you will find that logic easily and will need to change only this class

**Pull request merged**